README.md
# aliyundrive-webdav

[![GitHub Actions](https://github.com/messense/aliyundrive-webdav/workflows/CI/badge.svg)](https://github.com/messense/aliyundrive-webdav/actions?query=workflow%3ACI)
[![PyPI](https://img.shields.io/pypi/v/aliyundrive-webdav.svg)](https://pypi.org/project/aliyundrive-webdav)
[![Docker Image](https://img.shields.io/docker/pulls/messense/aliyundrive-webdav.svg?maxAge=2592000)](https://hub.docker.com/r/messense/aliyundrive-webdav/)
[![aliyundrive-webdav](https://snapcraft.io/aliyundrive-webdav/badge.svg)](https://snapcraft.io/aliyundrive-webdav)
[![Crates.io](https://img.shields.io/crates/v/aliyundrive-webdav.svg)](https://crates.io/crates/aliyundrive-webdav)

> 🚀 Help me to become a full-time open-source developer by [sponsoring me on GitHub](https://github.com/sponsors/messense)

阿里云盘 WebDAV 服务，主要使用场景为配合支持 WebDAV 协议的客户端 App 如 [Infuse](https://firecore.com/infuse)、[nPlayer](https://nplayer.com)
等实现在电视上直接观看云盘视频内容， 支持客户端 App 直接从阿里云盘获取文件播放而不经过运行本应用的服务器中转, 支持上传文件，但受限于 WebDAV 协议不支持文件秒传。

**请注意：V2 版本基于阿里云盘开放平台接口实现，不再支持阿里云盘 Web 和 App 版本获取的 refresh token。**
**由于本项目作者不再使用梅林固件，V2 版本不再免费支持 Koolshare 梅林固件系统，如有需要请考虑[付费支持](https://github.com/messense/aliyundrive-webdav/discussions/778)。**

如果项目对你有帮助，请考虑[捐赠支持](https://github.com/messense/aliyundrive-webdav/discussions/126)项目持续维护。
也可以考虑加入[aliyundrive-webdav 知识星球](https://t.zsxq.com/0c9sq6Ca8)获取咨询和技术支持服务。

> **Note**
>
> 本项目作者没有上传需求, 故上传文件功能测试不全面且没有持续迭代计划.

## 安装

可以从 [GitHub Releases](https://github.com/messense/aliyundrive-webdav/releases) 页面下载预先构建的二进制包， 也可以使用 pip 从 PyPI 下载:

```bash
pip install aliyundrive-webdav
```

如果系统支持 [Snapcraft](https://snapcraft.io) 比如 Ubuntu、Debian 等，也可以使用 snap 安装：

```bash
sudo snap install aliyundrive-webdav
```

### OpenWrt 路由器

[GitHub Releases](https://github.com/messense/aliyundrive-webdav/releases) 中有预编译的 ipk 文件， 目前提供了
aarch64/arm/mipsel/x86_64/i686 等架构的版本，可以下载后使用 opkg 安装，以 nanopi r4s 为例：

```bash
wget https://github.com/messense/aliyundrive-webdav/releases/download/v2.3.2/aliyundrive-webdav_2.3.2-1_aarch64_generic.ipk
wget https://github.com/messense/aliyundrive-webdav/releases/download/v2.3.2/luci-app-aliyundrive-webdav_2.3.2_all.ipk
wget https://github.com/messense/aliyundrive-webdav/releases/download/v2.3.2/luci-i18n-aliyundrive-webdav-zh-cn_2.3.2-1_all.ipk
opkg install aliyundrive-webdav_2.3.2-1_aarch64_generic.ipk
opkg install luci-app-aliyundrive-webdav_2.3.2_all.ipk
opkg install luci-i18n-aliyundrive-webdav-zh-cn_2.3.2-1_all.ipk
```

其它 CPU 架构的路由器可在 [GitHub Releases](https://github.com/messense/aliyundrive-webdav/releases) 页面中查找对应的架构的主程序 ipk 文件下载安装， 常见
OpenWrt 路由器 CPU 架构如下表（欢迎补充）：

| 路由器       | CPU 架构           |
| ------------ | ------------------ |
| nanopi r4s   | aarch64_generic    |
| 小米 AX3600  | aarch64_cortex-a53 |
| 斐讯 N1 盒子 | aarch64_cortex-a53 |
| Newifi D2    | mipsel_24kc        |
| Pogoplug     | arm_mpcore         |

> Tips: 不清楚 CPU 架构类型可通过运行 `opkg print-architecture` 命令查询。

![OpenWrt 配置界面](./doc/openwrt.png)

## Docker 运行

```bash
docker run -d --name=aliyundrive-webdav --restart=unless-stopped -p 8080:8080 \
  -v /etc/aliyundrive-webdav/:/etc/aliyundrive-webdav/ \
  -e REFRESH_TOKEN='your refresh token' \
  -e WEBDAV_AUTH_USER=admin \
  -e WEBDAV_AUTH_PASSWORD=admin \
  messense/aliyundrive-webdav
```

其中，`REFRESH_TOKEN` 环境变量为你的阿里云盘 `refresh_token`，`WEBDAV_AUTH_USER`
和 `WEBDAV_AUTH_PASSWORD` 为连接 WebDAV 服务的用户名和密码。

### QNAP 威联通 NAS

#### QNAP 插件

[qnap-aliyunpan-webdav](https://github.com/iranee/qnap-aliyunpan-webdav) by
[@iranee](https://github.com/iranee).

#### ContainerStation (Docker)

管理员登陆 NAS 后安装 ContainerStation 并启动服务，在 Management (管理) 标签中 Create Application (新建应用)，配置如下

```yaml
version: '3.3'
services:
  aliyundrive-webdav:
    container_name: aliyundrive-webdav
    restart: unless-stopped
    ports:
      - '8080:8080'
    environment:
      - 'REFRESH_TOKEN=mytoken...'
    image: messense/aliyundrive-webdav
```

其中 `REFRESH_TOKEN` 文档最下面说明；`:8080` 网盘访问映射端口，可以按需改为其他的。

点击 Create (创建)后启动，访问 http://nas地址:8080/ 即可看到你网盘的自动生成索引网页文件。

参考文档

- https://docs.docker.com/compose/
- https://www.composerize.com/

## rclone

为了避免重复上传文件，使用 rclone 时推荐使用 [Nextcloud WebDAV](https://rclone.org/webdav/#nextcloud) 模式，可以支持 sha1 checksums. 
另外需要配合 `--no-update-modtime` 参数，否则 rclone 为了更新文件修改时间还是会强制重新上传。

举个例子：

```bash
rclone --no-update-modtime copy abc.pdf aliyundrive-nc://docs/
```

## 获取 refresh token

* [通过在线工具获取 refresh token](https://messense-aliyundrive-webdav-backendrefresh-token-ucs0wn.streamlit.app/)
* 命令行运行 `aliyundrive-webdav qr login` 扫码授权后会输出 refresh token

![扫码获取 refresh token](./doc/refresh-token.png)

## 命令行用法

```bash
$ aliyundrive-webdav --help
WebDAV server for AliyunDrive

Usage: aliyundrive-webdav [OPTIONS]
       aliyundrive-webdav <COMMAND>

Commands:
  qr    Scan QRCode
  help  Print this message or the help of the given subcommand(s)

Options:
      --host <HOST>
          Listen host

          [env: HOST=]
          [default: 0.0.0.0]

  -p, --port <PORT>
          Listen port

          [env: PORT=]
          [default: 8080]

      --client-id <CLIENT_ID>
          Aliyun drive client_id

          [env: CLIENT_ID=]

      --client-secret <CLIENT_SECRET>
          Aliyun drive client_secret

          [env: CLIENT_SECRET=]

      --drive-type <DRIVE_TYPE>
          Aliyun drive type

          [env: DRIVE_TYPE=]

          Possible values:
          - resource: Resource drive
          - backup:   Backup drive
          - default:  Default drive

  -r, --refresh-token <REFRESH_TOKEN>
          Aliyun drive refresh token

          [env: REFRESH_TOKEN=]

  -U, --auth-user <AUTH_USER>
          WebDAV authentication username

          [env: WEBDAV_AUTH_USER=]

  -W, --auth-password <AUTH_PASSWORD>
          WebDAV authentication password

          [env: WEBDAV_AUTH_PASSWORD=]

  -I, --auto-index
          Automatically generate index.html

  -S, --read-buffer-size <READ_BUFFER_SIZE>
          Read/download buffer size in bytes, defaults to 10MB

          [default: 10485760]

      --upload-buffer-size <UPLOAD_BUFFER_SIZE>
          Upload buffer size in bytes, defaults to 16MB

          [default: 16777216]

      --cache-size <CACHE_SIZE>
          Directory entries cache size

          [default: 1000]

      --cache-ttl <CACHE_TTL>
          Directory entries cache expiration time in seconds

          [default: 600]

      --root <ROOT>
          Root directory path

          [default: /]

  -w, --workdir <WORKDIR>
          Working directory, refresh_token will be stored in there if specified

      --no-trash
          Delete file permanently instead of trashing it

      --read-only
          Enable read only mode

      --tls-cert <TLS_CERT>
          TLS certificate file path

          [env: TLS_CERT=]

      --tls-key <TLS_KEY>
          TLS private key file path

          [env: TLS_KEY=]

      --strip-prefix <STRIP_PREFIX>
          Prefix to be stripped off when handling request

          [env: WEBDAV_STRIP_PREFIX=]

      --debug
          Enable debug log

      --no-self-upgrade
          Disable self auto upgrade

      --skip-upload-same-size
          Skip uploading same size file

      --prefer-http-download
          Prefer downloading using HTTP protocol

      --redirect
          Enable 302 redirect when possible

  -h, --help
          Print help (see a summary with '-h')

  -V, --version
          Print version
```

> **Note**
> 
> 注意：TLS/HTTPS 暂不支持 MIPS 架构。

> **Note**
> 
> 注意：启用 `--skip-upload-same-size` 选项虽然能加速上传但可能会导致修改过的同样大小的文件不会被上传

## License

This work is released under the MIT license. A copy of the license is provided in the [LICENSE](./LICENSE) file.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 2.3.0

* 增加资源库支持

## 2.1.0

* 增加 `--redirect` 参数用于启用 302 重定向

## 2.0.0

* 切换到阿里云盘开放平台接口
* 移除 Koolshare 梅林固件路由器平台支持

## 1.11.0

* 移除阿里云 PDS 服务支持

## 1.10.1

* 修复使用 Web 版 refresh token 时下载被错误 302 重定向的问题

## 1.10.0

* 使用 App refresh token 下载时默认 302 重定向而不需要中转

## 1.9.0

* 增加使用 HTTP 协议下载配置,低端设备中转时降低资源消耗

## 1.8.9

* 修复上传大文件时上传地址过期的问题

## 1.8.8

* 修复开启 TLS 后清除缓存导致进程 crash 的问题

## 1.8.7

* 复制/删除文件夹时清除原文件夹缓存内容

## 1.8.6

* 修复重命名文件夹时原文件夹缓存内容未清除的问题

## 1.8.5

* 支持 rclone 以 Nextcloud WebDAV 模式上传时跳过上传相同 sha1 哈希值文件

## 1.8.4

* 支持 rclone 以 OwnCloud/Nextcloud WebDAV 模式挂载时返回 sha1 checksum

## 1.8.3

* 优化上传文件完成目录缓存失效策略

## 1.8.2

* 修复读取目录在阿里云盘接口请求错误时返回 404 的问题
* OpenWrt 界面增加清除缓存功能

## 1.8.1

* 增加调试模式 HTTP 请求日志输出

## 1.8.0

* 增加配置上传文件缓冲区大小参数 `--upload-buffer-size`
* 增加配置跳过上传相同大小同名文件参数 `--skip-upload-same-size`, 注意启用该选项虽然能加速上传但可能会导致修改过的同样大小的文件不会被上传

## 1.7.4

* 删除文件时忽略 404 和 400 状态码
* 修复梅林 arm384/arm386 使用 usb2jffs 插件后安装报错 `bad number` 问题
* 上传文件出错时日志中增加更详细的错误信息

## 1.7.3

* 调用云盘接口增加自动重试机制

## 1.7.2

* 增加 socks5 代理支持

## 1.7.1

* OpenWrt Luci 配置界面增加扫码登录获取 refresh token 功能

## 1.7.0

* 梅林 384/386 固件禁用程序自动更新
* 默认使用 App refresh token 刷新接口
* 增加 `aliyundrive-webdav qr` 子命令

## 1.6.2

* 非 tty 终端模式下不尝试扫码登录

## 1.6.1

* 降低自动更新失败日志级别为警告

## 1.6.0

* 增加自动更新功能

## 1.5.1

* 修复 Web 版 refresh token 刷新失败问题

## 1.5.0

* 增加移动端 App refresh token 支持,扫码登录使用 App refresh token.

## 1.4.0

* 命令行增加阿里云盘扫码登录功能

## 1.3.3

* 增加 `--strip-prefix` 参数

## 1.3.2

* 不使用阿里云盘文件列表接口返回的可能有问题的图片下载地址

## 1.3.1

* 降低获取文件下载地址接口调用次数

## 1.3.0

* 支持下载 `.livp` 格式文件

## 1.2.7

* 修复下载部分文件类型如 `.livp` 500 报错问题，由于阿里云盘接口没有返回 `.livp` 文件格式下载地址，暂时无法下载该格式文件

## 1.2.6

* 指定 `--workdir` 参数时 `--refresh-token` 参数可选

## 1.2.5

* 修复 Windows 版本访问文件 404 问题

## 1.2.4

* 修正 OpenWrt package autorelease 版本号

## 1.2.3

* 增加 Windows arm64 架构支持

## 1.2.2

* TLS/HTTPS 支持 RSA 私钥格式

## 1.2.1

* 支持 OpenWrt 19.07

## 1.2.0

* 增加 TLS/HTTPS 支持（暂不支持 MIPS 架构）
* 增加 HTTP 2.0 支持
* 修复 Docker 容器设置 `HOST` 环境变量不生效的问题
* 增加构建发布 deb 和 rpm 包

## 1.1.1

* 修复潜在的内存泄漏问题

## 1.1.0

* 增加只读模式，防止误操作删除文件

## 1.0.0

* 调整连接池 idle 检测时间，避免下载文件时出现 `connection closed before message
  completed` 报错
* 功能相对稳定，发布 1.0 版本。

## 0.5.5

* 降级 OpenSSL 修复 MIPS 架构二进制文件无法正常运行的问题

## 0.5.4

* 刷新 refresh token 增加 429 状态码重试

## 0.5.3

* 完善请求重试，处理请求 408、429 报错

## 0.5.2

* 增加 `arm_cortex-a5_vfpv4` 架构 OpenWrt 包（玩客云适用）

## 0.5.1

* 修复 OpenWrt Luci 界面语言翻译问题

## 0.5.0

* 增加实验性[阿里云相册与网盘服务（PDS）](https://www.aliyun.com/product/storage/pds)支持，阿里云网站开通 PDS 服务后可通过传入 `domain_id` 和对应用户的 `refresh_token`（可通过访问 BasicUI 获取） 使用。

## 0.4.8

* 支持通过环境变量 `HOST` 和 `PORT` 配置监听地址和端口

## 0.4.7

* 发布 musllinux wheel 二进制包至 PyPI

## 0.4.6

* 自动尝试刷新过期的上传地址
* GitHub Release 产物文件名增加版本号

## 0.4.5

* 兼容 macOS Finder chunked encoding 上传 `X-Expected-Entity-Length` HTTP header

## 0.4.4

* 新增目录缓存过期时间参数配置


src/main.rs
use std::env;
use std::path::PathBuf;

use anyhow::bail;
use clap::{Parser, Subcommand};
use dav_server::{memls::MemLs, DavHandler};
#[cfg(unix)]
use futures_util::stream::StreamExt;
use self_update::cargo_crate_version;
use tracing::{debug, info, warn};
#[cfg(unix)]
use {signal_hook::consts::signal::*, signal_hook_tokio::Signals};

use cache::Cache;
use drive::{read_refresh_token, AliyunDrive, DriveConfig, DriveType};
use vfs::AliyunDriveFileSystem;
use webdav::WebDavServer;

mod cache;
mod drive;
mod login;
mod vfs;
mod webdav;

#[derive(Parser, Debug)]
#[command(name = "aliyundrive-webdav", about, version, author)]
#[command(args_conflicts_with_subcommands = true)]
struct Opt {
    /// Listen host
    #[arg(long, env = "HOST", default_value = "0.0.0.0")]
    host: String,
    /// Listen port
    #[arg(short, env = "PORT", long, default_value = "8080")]
    port: u16,
    /// Aliyun drive client_id
    #[arg(long, env = "CLIENT_ID")]
    client_id: Option<String>,
    /// Aliyun drive client_secret
    #[arg(long, env = "CLIENT_SECRET")]
    client_secret: Option<String>,
    /// Aliyun drive type
    #[arg(long, env = "DRIVE_TYPE")]
    drive_type: Option<DriveType>,
    /// Aliyun drive refresh token
    #[arg(short, long, env = "REFRESH_TOKEN")]
    refresh_token: Option<String>,
    /// WebDAV authentication username
    #[arg(short = 'U', long, env = "WEBDAV_AUTH_USER")]
    auth_user: Option<String>,
    /// WebDAV authentication password
    #[arg(short = 'W', long, env = "WEBDAV_AUTH_PASSWORD")]
    auth_password: Option<String>,
    /// Automatically generate index.html
    #[arg(short = 'I', long)]
    auto_index: bool,
    /// Read/download buffer size in bytes, defaults to 10MB
    #[arg(short = 'S', long, default_value = "10485760")]
    read_buffer_size: usize,
    /// Upload buffer size in bytes, defaults to 16MB
    #[arg(long, default_value = "16777216")]
    upload_buffer_size: usize,
    /// Directory entries cache size
    #[arg(long, default_value = "1000")]
    cache_size: u64,
    /// Directory entries cache expiration time in seconds
    #[arg(long, default_value = "600")]
    cache_ttl: u64,
    /// Root directory path
    #[arg(long, default_value = "/")]
    root: String,
    /// Working directory, refresh_token will be stored in there if specified
    #[arg(short = 'w', long)]
    workdir: Option<PathBuf>,
    /// Delete file permanently instead of trashing it
    #[arg(long)]
    no_trash: bool,
    /// Enable read only mode
    #[arg(long)]
    read_only: bool,
    /// TLS certificate file path
    #[arg(long, env = "TLS_CERT")]
    tls_cert: Option<PathBuf>,
    /// TLS private key file path
    #[arg(long, env = "TLS_KEY")]
    tls_key: Option<PathBuf>,
    /// Prefix to be stripped off when handling request.
    #[arg(long, env = "WEBDAV_STRIP_PREFIX")]
    strip_prefix: Option<String>,
    /// Enable debug log
    #[arg(long)]
    debug: bool,
    /// Disable self auto upgrade
    #[arg(long)]
    no_self_upgrade: bool,
    /// Skip uploading same size file
    #[arg(long)]
    skip_upload_same_size: bool,
    /// Prefer downloading using HTTP protocol
    #[arg(long)]
    prefer_http_download: bool,
    /// Enable 302 redirect when possible
    #[arg(long)]
    redirect: bool,

    #[command(subcommand)]
    subcommands: Option<Commands>,
}

#[derive(Subcommand, Debug)]
enum Commands {
    /// Scan QRCode
    #[command(subcommand)]
    Qr(QrCommand),
}

#[derive(Subcommand, Debug)]
enum QrCommand {
    /// Scan QRCode login to get a token
    Login,
    /// Generate a QRCode
    Generate,
    /// Query the QRCode login result
    #[command(arg_required_else_help = true)]
    Query {
        /// Query parameter sid
        #[arg(long)]
        sid: String,
    },
}

#[tokio::main(flavor = "multi_thread")]
async fn main() -> anyhow::Result<()> {
    #[cfg(feature = "native-tls-vendored")]
    openssl_probe::init_ssl_cert_env_vars();

    let opt = Opt::parse();
    if env::var("RUST_LOG").is_err() {
        if opt.debug {
            env::set_var("RUST_LOG", "aliyundrive_webdav=debug,reqwest=debug");
        } else {
            env::set_var("RUST_LOG", "aliyundrive_webdav=info,reqwest=warn");
        }
    }
    tracing_subscriber::fmt()
        .with_timer(tracing_subscriber::fmt::time::LocalTime::rfc_3339())
        .init();

    let workdir = opt
        .workdir
        .or_else(|| dirs::cache_dir().map(|c| c.join("aliyundrive-webdav")));
    let refresh_token_host = if opt.client_id.is_none() || opt.client_secret.is_none() {
        env::var("ALIYUNDRIVE_OAUTH_SERVER")
            .unwrap_or_else(|_| "https://aliyundrive-oauth.messense.me".to_string())
    } else {
        "https://openapi.aliyundrive.com".to_string()
    };
    let drive_config = DriveConfig {
        api_base_url: "https://openapi.aliyundrive.com".to_string(),
        refresh_token_host,
        workdir,
        client_id: opt.client_id.clone(),
        client_secret: opt.client_secret.clone(),
        drive_type: opt.drive_type.clone(),
    };

    // subcommands
    if let Some(Commands::Qr(qr)) = opt.subcommands.as_ref() {
        match qr {
            QrCommand::Login => {
                let refresh_token = login(drive_config.clone(), 120).await?;
                println!("\nrefresh_token:\n\n{}", refresh_token)
            }
            QrCommand::Generate => {
                let scanner = login::QrCodeScanner::new(drive_config.clone()).await?;
                let data = scanner.scan().await?;
                println!("{}", serde_json::to_string_pretty(&data)?);
            }
            QrCommand::Query { sid } => {
                let scanner = login::QrCodeScanner::new(drive_config.clone()).await?;
                let query_result = scanner.query(sid).await?;
                if query_result.is_success() {
                    let code = query_result.auth_code.unwrap();
                    let refresh_token = scanner.fetch_refresh_token(&code).await?;
                    println!("{}", refresh_token)
                }
            }
        }
        return Ok(());
    }

    if env::var("NO_SELF_UPGRADE").is_err() && !opt.no_self_upgrade {
        tokio::task::spawn_blocking(move || {
            if let Err(e) = check_for_update(opt.debug) {
                debug!("failed to check for update: {}", e);
            }
        })
        .await?;
    }

    let auth_user = opt.auth_user;
    let auth_password = opt.auth_password;
    if (auth_user.is_some() && auth_password.is_none())
        || (auth_user.is_none() && auth_password.is_some())
    {
        bail!("auth-user and auth-password must be specified together.");
    }

    let tls_config = match (opt.tls_cert, opt.tls_key) {
        (Some(cert), Some(key)) => Some((cert, key)),
        (None, None) => None,
        _ => bail!("tls-cert and tls-key must be specified together."),
    };

    let refresh_token_from_file = if let Some(dir) = drive_config.workdir.as_ref() {
        read_refresh_token(dir).await.ok()
    } else {
        None
    };
    let refresh_token = if opt.refresh_token.is_none()
        && refresh_token_from_file.is_none()
        && atty::is(atty::Stream::Stdout)
    {
        login(drive_config.clone(), 30).await?
    } else {
        let token = opt.refresh_token.unwrap_or_default();
        if !token.is_empty() && token.split('.').count() < 3 {
            bail!("Invalid refresh token value found in `--refresh-token` argument");
        }
        token
    };

    let drive = AliyunDrive::new(drive_config, refresh_token).await?;
    let mut fs = AliyunDriveFileSystem::new(drive, opt.root, opt.cache_size, opt.cache_ttl)?;
    fs.set_no_trash(opt.no_trash)
        .set_read_only(opt.read_only)
        .set_upload_buffer_size(opt.upload_buffer_size)
        .set_skip_upload_same_size(opt.skip_upload_same_size)
        .set_prefer_http_download(opt.prefer_http_download);
    debug!("aliyundrive file system initialized");

    #[cfg(unix)]
    let dir_cache = fs.dir_cache.clone();

    let mut dav_server_builder = DavHandler::builder()
        .filesystem(Box::new(fs))
        .locksystem(MemLs::new())
        .read_buf_size(opt.read_buffer_size)
        .autoindex(opt.auto_index)
        .redirect(opt.redirect);
    if let Some(prefix) = opt.strip_prefix {
        dav_server_builder = dav_server_builder.strip_prefix(prefix);
    }

    let dav_server = dav_server_builder.build_handler();
    debug!(
        read_buffer_size = opt.read_buffer_size,
        auto_index = opt.auto_index,
        "webdav handler initialized"
    );

    let server = WebDavServer {
        host: opt.host,
        port: opt.port,
        auth_user,
        auth_password,
        tls_config,
        handler: dav_server,
    };

    #[cfg(not(unix))]
    server.serve().await?;

    #[cfg(unix)]
    {
        let signals = Signals::new([SIGHUP])?;
        let handle = signals.handle();
        let signals_task = tokio::spawn(handle_signals(signals, dir_cache));

        server.serve().await?;

        // Terminate the signal stream.
        handle.close();
        signals_task.await?;
    }
    Ok(())
}

#[cfg(unix)]
async fn handle_signals(mut signals: Signals, dir_cache: Cache) {
    while let Some(signal) = signals.next().await {
        match signal {
            SIGHUP => {
                dir_cache.invalidate_all();
                info!("directory cache invalidated by SIGHUP");
            }
            _ => unreachable!(),
        }
    }
}

async fn login(drive_config: DriveConfig, timeout: u64) -> anyhow::Result<String> {
    const SLEEP: u64 = 3;

    let scanner = login::QrCodeScanner::new(drive_config).await?;
    // 返回二维码内容结果集
    let sid = scanner.scan().await?.sid;
    // 需要生成二维码的内容
    let qrcode_content = format!("https://www.aliyundrive.com/o/oauth/authorize?sid={sid}");
    // 打印二维码
    qr2term::print_qr(&qrcode_content)?;
    info!("Please scan the qrcode to login in {} seconds", timeout);
    let loop_count = timeout / SLEEP;
    for _i in 0..loop_count {
        tokio::time::sleep(tokio::time::Duration::from_secs(SLEEP)).await;
        // 模拟轮训查询二维码状态
        let query_result = scanner.query(&sid).await?;
        if !query_result.is_success() {
            continue;
        }
        let code = query_result.auth_code.unwrap();
        let refresh_token = scanner.fetch_refresh_token(&code).await?;
        return Ok(refresh_token);
    }
    bail!("Login failed")
}

fn check_for_update(show_output: bool) -> anyhow::Result<()> {
    use self_update::update::UpdateStatus;
    #[cfg(unix)]
    use std::os::unix::process::CommandExt;
    use std::process::Command;

    let auth_token = env::var("GITHUB_TOKEN")
        .unwrap_or_else(|_| env::var("HOMEBREW_GITHUB_API_TOKEN").unwrap_or_default());
    let status = self_update::backends::github::Update::configure()
        .repo_owner("messense")
        .repo_name("aliyundrive-webdav")
        .bin_name("aliyundrive-webdav")
        .target(if cfg!(target_os = "macos") {
            "apple-darwin"
        } else {
            self_update::get_target()
        })
        .auth_token(&auth_token)
        .show_output(show_output)
        .show_download_progress(true)
        .no_confirm(true)
        .current_version(cargo_crate_version!())
        .build()?
        .update_extended()?;
    if let UpdateStatus::Updated(ref release) = status {
        if let Some(body) = &release.body {
            if !body.trim().is_empty() {
                info!("aliyundrive-webdav upgraded to {}:\n", release.version);
                info!("{}", body);
            } else {
                info!("aliyundrive-webdav upgraded to {}", release.version);
            }
        }
    } else {
        info!("aliyundrive-webdav is up-to-date");
    }

    if status.updated() {
        warn!("Respawning...");
        let current_exe = env::current_exe();
        let mut command = Command::new(current_exe?);
        command.args(env::args().skip(1)).env("NO_SELF_UPGRADE", "");
        #[cfg(unix)]
        {
            let err = command.exec();
            bail!(err);
        }

        #[cfg(windows)]
        {
            let status = command.spawn().and_then(|mut c| c.wait())?;
            bail!("aliyundrive-webdav upgraded");
        }
    }
    Ok(())
}
